library/objective-c/Envoy.mm

```diff
@@ -0,0 +1,45 @@
+#import "library/objective-c/Envoy.h"
+
+#import "library/common/main_interface.h"
+
+
+@interface Envoy ()
+@property (nonatomic, strong) NSThread *runner;
+@end
+
+@implementation Envoy
+
+@synthesize runner;
+
+// Create a new Envoy instance. The Envoy runner NSThread is started as part of instance
+// initialization with the configuration provided.
+- (id)initWithConfig:(NSString *)config {
+  self.runner = [[NSThread alloc] initWithTarget:self selector:@selector(run:) object:config];
+  [self.runner start];
+}
+
+// Returns whether this Envoy instance is currently active and running.
+- (BOOL)isRunning {
+  return self.runner != NULL && self.runner.isExecuting;
+}
+
+// Returns whether the Envoy instance is terminated.
+- (BOOL)isTerminated {
+  return self.runner != NULL && self.runner.isFinished;
+}
+
+#pragma mark private
+
+- (void)run:(NSString *)config {
+  try {
+    run_envoy(config.UTF8String);
+  } catch (NSException *e) {
+    NSLog(@"Envoy exception: %@", e);
+    NSDictionary *userInfo = @{ @"exception": e};
+    [NSNotificationCenter.defaultCenter postNotificationName:@"EnvoyException"
```

Comment:
Do you think we should prefix this notification name with something to better differentiate? Would also be nice to document it somewhere

Reply:
If we think this is good pattern (and I think maybe it's a useful one), then yeah we should document.

What sort of prefix did you have in mind?

Reply:
(I was sort of thinking of "Envoy" as a prefix fwiw)

Reply:
Objective-C normally asks for 3-letter class prefixes, so maybe we could do the same for notifications 🤔

Reply:
How strict is that convention these days? My inclination is to eschew it, unless it's really going to bother people.

Reply:
It can still cause collisions in Objective-C classes 🤷‍♂

Reply:
Though, notifications might also be fine

Reply:
I don't care that much, but I can't think of a good letter code that's not going to risk just making things more confusing. "Envoy" as a prefix seems (to me) pretty low-risk in terms of likelihood to cause a collision.

Reply:
Fine with me to leave this as-is for now

Reply:
👍